City-of-Helsinki/hel-fi-drupal-grants
# feat: AU-2105: Data validation for LIIKUNTATILANKAYTTO 60. Possibility to skip validation. validateFormErrors fix.

# [AU-2105](https://helsinkisolutionoffice.atlassian.net/browse/AU-2105)

## What was done

This pull requests:
- Adds validation to the form `LIIKUNTATILANKAYTTO 60`. 

- Fixes the function `validateFormErrors` so that it actually parses and validates all errors present in an `expectedErrors` object.

- Adds a `viewPageSkipValidation` key to the `FormData` interface. This key can be set to `true` in all `missingValues` and `wrongValues` forms. If the value is set to true, then the form variant will not be validated. Said key has been added to all present forms and their tests have been modified accordingly.

- Changes the behavior of the `viewPageBuildSelectorForItem` function that's used to construct classes for form elements when performing validation. Now, the function only removes numbers from the end of an elements key (previously all numbers were removed). This makes the function return correct classes for keys like `edit-joista-helsinkilaisia-pojat-20-aktiiviharrastajat` and `edit-muut-20-aktiiviharrastajat`.

- Adds code to the `RentIncomeComposite.php` file so that the outputted dates in the composite element are formatted correctly.

## How to install

Make sure your instance is up and running on correct branch.
 * `git checkout feature/AU-2105-e2e-validation-60`
 * `make fresh`
 * `make drush-cr`

## How to test

- Should your tests fail, change the `.then()` function after `fetchLatestProfileByType` so that it returns `false`. This forces the user data to refresh.

- As I write this, no submitted forms are going through to Avus2. That's why the `success` key in `registeredCommunityApplications_60` is commented out.

- Run `make test-pw registered_community_60`

Make sure you get an output that looks like this:

![Screenshot 2024-01-31 at 16 44 18](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/fa7dc9b0-4f2a-4b51-8e9a-4521014a3615)


[AU-2105]: https://helsinkisolutionoffice.atlassian.net/browse/AU-2105?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
